ngrinder-controller/src/main/java/org/ngrinder/agent/service/AgentService.java

```diff
@@ -321,17 +325,47 @@ public synchronized void runAgent(User user, final SingleConsole singleConsole,
 	/**
 	 * Select agent. This method return agent set which is belong to the given user first and then share agent set.
 	 *
+	 * Priority of agent selection.
+	 * 1. owned agent of recently used.
+	 * 2. owned agent.
+	 * 3. public agent of recently used.
+	 * 4. public agent.
+	 *
 	 * @param user          user
-	 * @param allFreeAgents agents
-	 * @param agentCount    number of agent
-	 * @return selected agent.
+	 * @param allFreeAgents available agents
+	 * @param agentCount    number of agents
+	 * @return selected agents.
 	 */
-	private Set<AgentInfo> selectAgent(User user, Set<AgentInfo> allFreeAgents, int agentCount) {
-		Stream<AgentInfo> ownedFreeAgentStream = allFreeAgents.stream().filter(agentInfo -> isOwnedAgent(agentInfo, user.getUserId()));
-		Stream<AgentInfo> freeAgentStream = allFreeAgents.stream().filter(this::isCommonAgent);
+	Set<AgentInfo> selectAgent(User user, Set<AgentInfo> allFreeAgents, int agentCount) {
+		Set<AgentInfo> recentlyUsedAgents = hazelcastService.getOrDefault(CACHE_RECENTLY_USED_AGENTS, user.getUserId(), emptySet());
```

Comment:
How about using stream style like before?
Get `recentlyUsedAgents` as a stream and then, concat these all three streams and limit by `agentCount`.

Reply:
Yes, It better.
It is modified.